### DISCUSSION TITLE
I have a JSON of TRON and I want to sign it. What should I do

### DISCUSSION THREAD

**Discussion by tangang199354** [trustwallet/wallet-core#1436] [Q&A]
{
    "visible":false,
    "txID":"43e61ed5157578c658159f2fa741289e2c51031d26fd310502f97044617d8511",
    "raw_data":{
        "contract":[
            {
                "parameter":{
                    "value":{
                        "data":"4bf3e2d000000000000000000000000000000000000000000000000000000000012eafda00000000000000000000000000000000000000000000000000000000609c8241",
                        "owner_address":"411e6992467173863e666db86a6ebc84d2dffd0e05",
                        "contract_address":"414d750a57d1cadae660f911bf120a4e11a9f13bbb",
                        "call_value":1000000
                    },
                    "type_url":"type.googleapis.com/protocol.TriggerSmartContract"
                },
                "type":"TriggerSmartContract"
            }
        ],
        "ref_block_bytes":"e816",
        "ref_block_hash":"ecd82adc1d718100",
        "expiration":1620869691000,
        "fee_limit":40000000,
        "timestamp":1620869637327
    },
    "raw_data_hex":"0a02e8162208ecd82adc1d71810040f8ecf29a962f5ab201081f12ad010a31747970652e676f6f676c65617069732e636f6d2f70726f746f636f6c2e54726967676572536d617274436f6e747261637412780a15411e6992467173863e666db86a6ebc84d2dffd0e051215414d750a57d1cadae660f911bf120a4e11a9f13bbb18c0843d22444bf3e2d000000000000000000000000000000000000000000000000000000000012eafda00000000000000000000000000000000000000000000000000000000609c824170cfc9ef9a962f900180b48913",
    "payInfo":{
        "contract_address":"414d750a57d1cadae660f911bf120a4e11a9f13bbb",
        "owner_address":"411e6992467173863e666db86a6ebc84d2dffd0e05",
        "function_selector":"trxToTokenSwapInput(uint256,uint256)",
        "parameter":"00000000000000000000000000000000000000000000000000000000012eafda00000000000000000000000000000000000000000000000000000000609c8241",
        "call_value":1000000,
        "fee_limit":40000000
    }
}

**Comment by tangang199354** [comment]
I want to sign the JSON for this transfer, but I can't find a Function in the SDK that can sign it

**Reply by jiuyaopinyin** [reply]
hi,how did you finish it?

**Comment by optout21** [comment]
You have to sign a SigningInput object, and then invoke sign().
For examples, see tests:
https://github.com/trustwallet/wallet-core/blob/master/tests/Tron/TWAnySignerTests.cpp
https://github.com/trustwallet/wallet-core/blob/master/tests/Tron/SignerTests.cpp

**Reply by wcj233** [reply]
Hi @catenocrypt , I have the same issue. I tried to create a SigningInput object, and then invoke sign(), at last, when I called the broadcasttransaction api, it returned "SIGERROR", I didn't know what step was wrong, could you help me? Thanks!

The data of Tron:
```
▿ 4 elements
  ▿ 0 : 2 elements
    - key : "txID"
    - value : 9e17d5b4929eb962540a1bf00686b450f2d22f3c29f2c0cbbe9338048086e60e
  ▿ 1 : 2 elements
    - key : "raw_data_hex"
    - value : 0a02f7192208c9922c47b15a383e4088e3cac1bf2f5ab101081f12ac010a31747970652e676f6f676c65617069732e636f6d2f70726f746f636f6c2e54726967676572536d617274436f6e747261637412770a1541f4ea52c222af5033009f9c0eac9b31e78174658c121541ec5e8661f8855b1724d4f668f11ab722b4b3b74d18904e22444bf3e2d0000000000000000000000000000000000000000000000000000000000000042b000000000000000000000000000000000000000000000000000000006145aec970e9a9c7c1bf2f900180e1eb17
  ▿ 2 : 2 elements
    - key : "visible"
    - value : 0
  ▿ 3 : 2 elements
    - key : "raw_data"
    ▿ value : 6 elements
      ▿ 0 : 2 elements
        - key : contract
        ▿ value : 1 element
          ▿ 0 : 2 elements
            ▿ 0 : 2 elements
              - key : type
              - value : TriggerSmartContract
            ▿ 1 : 2 elements
              - key : parameter
              ▿ value : 2 elements
                ▿ 0 : 2 elements
                  - key : value
                  ▿ value : 4 elements
                    ▿ 0 : 2 elements
                      - key : contract_address
                      - value : 41ec5e8661f8855b1724d4f668f11ab722b4b3b74d
                    ▿ 1 : 2 elements
                      - key : data
                      - value : 4bf3e2d0000000000000000000000000000000000000000000000000000000000000042b000000000000000000000000000000000000000000000000000000006145aec9
                    ▿ 2 : 2 elements
                      - key : call_value
                      - value : 10000
                    ▿ 3 : 2 elements
                      - key : owner_address
                      - value : 41f4ea52c222af5033009f9c0eac9b31e78174658c
                ▿ 1 : 2 elements
                  - key : type_url
                  - value : type.googleapis.com/protocol.TriggerSmartContract
      ▿ 1 : 2 elements
        - key : ref_block_bytes
        - value : f719
      ▿ 2 : 2 elements
        - key : timestamp
        - value : 1631956620521
      ▿ 3 : 2 elements
        - key : fee_limit
        - value : 50000000
      ▿ 4 : 2 elements
        - key : expiration
        - value : 1631956677000
      ▿ 5 : 2 elements
        - key : ref_block_hash
        - value : c9922c47b15a383e
```

I tried to create a SigningInput object:

```
let raw_data = dataDic["raw_data"] as! Dictionary<String, Any>
let contract = raw_data["contract"] as! Array<Any>
let firstDic = contract.first as! Dictionary<String, Any>
let parmater = firstDic["parameter"] as! Dictionary<String, Any>
let value = parmater["value"] as! Dictionary<String, Any>
let signInput = TronSigningInput.with {
            $0.transaction = TronTransaction.with {
                $0.blockHeader = TronBlockHeader.with {
                    $0.timestamp = Int64(truncating: headRawDic["timestamp"] as! NSNumber)
                    $0.number = Int64(truncating: headRawDic["number"] as! NSNumber)
                    $0.parentHash = Data(hexString: (headRawDic["parentHash"] as! String))!
                    $0.txTrieRoot = Data(hexString: (headRawDic["txTrieRoot"] as! String))!
                    $0.witnessAddress = Data(hexString: (headRawDic["witness_address"] as! String))!
                    $0.version = Int32(truncating: headRawDic["version"] as! NSNumber)
                }
                
                $0.triggerSmartContract = TronTriggerSmartContract.with {
                    $0.ownerAddress = value["owner_address"] as! String
                    $0.contractAddress = value["contract_address"] as! String
                    $0.callValue = (value["call_value"] as! NSNumber).int64Value
                    $0.data = Data(hex: (value["data"] as! String))
                }
                
            }
            $0.privateKey = pk.data
        }
        
let msg: TronSigningOutput = AnySigner.sign(input: signInput, coin: .tron)    
```
